Splits CountRegistrationsByIP to exact-match and by /48.

cpu opened the pull request:
Prior to this PR the SA's `CountRegistrationsByIP` treated IPv6
differently than IPv4 by counting registrations within a /48 for IPv6 as
opposed to exact matches for IPv4. This PR updates
`CountRegistrationsByIP` to treat IPv4 and IPv6 the
same, always matching exactly. The existing RegistrationsPerIP rate
limit policy will be applied against this exact matching count.

A new `CountRegistrationsByIPRange` function is added to the SA that
performs the historic matching process, e.g. for IPv4 it counts exactly
the same as `CountRegistrationsByIP`, but for IPv6 it counts within
a /48.

A new `RegistrationsPerIPRange` rate limit policy is added to allow
configuring the threshold/window for the fuzzy /48 matching registration
limit. Stats for the "Exceeded" and "Pass" events for this rate limit are
separated into a separate `RegistrationsByIPRange` stats scope under
the `RateLimit` scope to allow us to track it separate from the exact 
registrations per IP rate limit.

Resolves https://github.com/letsencrypt/boulder/issues/2738

jsha commented on grpc/sa-wrappers.go:
```diff
@@ -610,6 +632,20 @@ func (sas StorageAuthorityServerWrapper) CountRegistrationsByIP(ctx context.Cont
+		return nil, errIncompleteRequest
+	}
+
+	count, err := sas.inner.CountRegistrationsByIPRange(ctx, net.IP(request.Ip), time.Unix(0, *request.Range.Earliest), time.Unix(0, *request.Range.Latest))
```
Let's format this as argument-per-line to make it easier to read.

jsha commented on ra/ra.go:
```diff
@@ -242,22 +242,37 @@ type certificateRequestEvent struct {
 
-func (ra *RegistrationAuthorityImpl) checkRegistrationLimit(ctx context.Context, ip net.IP) error {
-	limit := ra.rlPolicies.RegistrationsPerIP()
+func (ra *RegistrationAuthorityImpl) checkRegistrationLimits(ctx context.Context, ip net.IP) error {
```
It looks like we could remove a lot of duplication by creating a function that takes a rate limit policy and an SA RPC function, and calls the given RPC to check against the given policy. What do you think?

cpu replied:
Sure - I agree there's excessive duplication here. I'll rework.

jsha commented on features/features.go:
```diff
@@ -41,6 +42,7 @@ var features = map[FeatureFlag]bool{
 	RandomDirectoryEntry:     false,
 	IPv6First:                false,
 	DirectoryMeta:            false,
+	RegistrationsPerIPRange:  false,
```
The flag-gating here is actually not needed because rate limit policies have their own enablement flag. Specifically, if there is no section in the config for a given policy, rlPolicy.Enabled() will return false. So we can deploy this without a change to rate-limit-policies.yml.

It's worth noting a subtle thing: While this meets the deployability requirements (can be deployed without concurrent config changes), it actually introduces a change in behavior immediately on deploy: instead of querying for, and rate limiting on, the /48, it will immediately start doing an exact query, and rate limiting on that. I think you are correct here in making the judgement call that this small change in behavior is safe enough to take effect immediately upon deploy.

cpu replied:
> 
The flag-gating here is actually not needed because rate limit policies have their own enablement flag. Specifically, if there is no section in the config for a given policy, rlPolicy.Enabled() will return false. So we can deploy this without a change to rate-limit-policies.yml.

Ahh!! That makes sense. I think this is the first rate limit I've added and it didn't occur to me.

> It's worth noting a subtle thing: While this meets the deployability requirements (can be deployed without concurrent config changes), it actually introduces a change in behavior immediately on deploy: instead of querying for, and rate limiting on, the /48, it will immediately start doing an exact query, and rate limiting on that. I think you are correct here in making the judgement call that this small change in behavior is safe enough to take effect immediately upon deploy.

Another good point. If I was going to keep the feature flag I should have used it to prevent the /48 -> to exact change from taking place unless the flag were set.

I think given we're both comfortable with the change taking place immediately upon deploy & using the standard rate limit enabling to add the /48 rate I'll remove the feature flag.

Thanks for the detailed review~!

cpu commented:
@jsha Thanks again - Ready for another :mag:. I also noticed I hadn't separated out the stats collection to let us know which of the two registrations per IP rate limits were failing/succeeding. I also fixed that and updated the PR description.

jsha commented on ra/ra.go:
```diff
@@ -242,37 +244,65 @@ type certificateRequestEvent struct {
+
+// checkRegistrationIPLimit checks a specific registraton limit by using the
+// provided registrationCounter function to determine if the limit has been
+// exceeded for a given IP
```
*or IP range

rolandshoemaker commented on ra/ra.go:
```diff
@@ -242,22 +244,65 @@ type certificateRequestEvent struct {
+	}
+	ra.regByIPStats.Inc("Pass", 1)
+
+	// Check the registrations per IP range limit using the
```
Is there a reason not to short-circuit and return early if `ip` is a v4 address? It seems like this will just double the query rate for IPv4 addresses, which I presume are the majority we see, which doesn't seem ideal given DB performance problems we've seen in the past.

cpu replied:
I thought potentially we might want to add an IPv4 range limit in the future, but that's an extremely weak justification for increasing DB load. I agree with you - I think this can be updated to ret early. Thanks!

cpu replied:
Fixed in https://github.com/letsencrypt/boulder/pull/2782/commits/db2973789b64e9226198cf4ced871fdeaacc1bc7

cpu commented:
@rolandshoemaker I broke my test with my quick fix :crying_cat_face: - I'll have to address your feedback tomorrow morn, out of time for the day. Thanks for the :mag: !

cpu commented:
> I broke my test with my quick fix :crying_cat_face:

I read the docs on `ip.To4()` backwards :-) All fixed in 2f4a47b and I left a comment to help avoid confusion down the road.

@rolandshoemaker Ready for another :mag: - thanks!
